### ISSUE TITLE
Accessing program variable in the randomx_vm class

### ISSUE THREAD

**Issue by moneroexamples** [tevador/RandomX#44]
`alignas(64) randomx::Program program;` is `protected` member of `class randomx_vm`:

https://github.com/tevador/RandomX/blob/master/src/virtual_machine.hpp#L54

I don't see any way of accessing it from the derived classes. This makes it impossible to inspect the RandomX code used in generating POW in ` randomx_calculate_hash`.

https://github.com/tevador/RandomX/blob/master/src/randomx.cpp#L236

I think such ability is important and I was wondering if the `program` variable could be exposed. For example, by adding `getProgram` into `class randomx_vm`, in the following, or similar form:

```
 randomx::Program const* getProgram()
 {
       return &program;
 };
```

**Comment by tevador** [comment]
The RandomX API provides `randomx_vm*` as an opaque pointer and it's not intended for its members to be directly accessible.

But you can already inspect the generated programs using the provided `code-generator` executable, which is built by default together with the `benchmark`.

https://github.com/tevador/RandomX/blob/master/src/tests/code-generator.cpp

```
Usage: bin/code-generator [OPTIONS]
Supported options:
  --softAes         use software AES (default: x86 AES-NI)
  --nonce  N        seed nonce (default: 1000)
  --genAsm          generate x86-64 asm code for nonce N
  --genNative       generate RandomX code for nonce N
  --genSuperscalar  generate superscalar program for nonce N
```

**Comment by moneroexamples** [comment]
Thanks I used before that, but my aim a bit different. 

I want to take existing block mined on private v12 chain, and recreate randomx program used to calculate its POW.

I am able to re-calculated POW value/hash of said block, but so far am not able to recreate the program stored in in the instance of `randomx_vm` used to calculate the POW (I added getProgram in my local copy of randomx).

My understanding was that if I use the code snippet from below with the `hash/seed` obtained from the existing block I will get same program as the one in `randomx_vm`: 
https://github.com/tevador/RandomX/blob/378d5def38e72ca391feb68905559a98dabce199/src/tests/code-generator.cpp#L69-L77

The hash/seed seems to be correct since I recreate POW value, but the program generated is different. At the moment not sure why.

**Comment by tevador** [comment]
Each hash uses 8 different programs and the code of each program depends on the output of the program before it.

So the code snippet from the code-generator will allow you to examine only the first program. If you want to see all 8 programs, you have to also execute at least the first 7 programs.

The best way to do that is, as you suggested, in the `randomx_calculate_hash` function.

Add this method to the `random_vm` class:

```c++
const randomx::Program& getProgram()
{
   return program;
}
```
And then modify the `randomx_calculate_hash` function like this:

```c++
void randomx_calculate_hash(randomx_vm *machine, const void *input, size_t inputSize, void *output) {
  alignas(16) uint64_t tempHash[8];
  blake2b(tempHash, sizeof(tempHash), input, inputSize, nullptr, 0);
  machine->initScratchpad(&tempHash);
  machine->resetRoundingMode();
  for (int chain = 0; chain < RANDOMX_PROGRAM_COUNT - 1; ++chain) {
    machine->run(&tempHash);
    std::cout << machine->getProgram() << std::endl; //print programs 0-6
    blake2b(tempHash, sizeof(tempHash), machine->getRegisterFile(), sizeof(randomx::RegisterFile), nullptr, 0);
  }
  machine->run(&tempHash);
  std::cout << machine->getProgram() << std::endl; //print program 7
  machine->getFinalResult(output, RANDOMX_HASH_SIZE);
}
```

Use the JIT compiled VM:
```c++
randomx_vm* machine = randomx_create_vm(RANDOMX_FLAG_JIT, cache, nullptr);
randomx_calculate_hash(machine, blob, sizeof(blob), output);
```

**Comment by moneroexamples** [comment]
Yes, this is exactly what I did in my local clone of the randomx. The issue is that it requires patching randomx for anyone who would want to replicate this. 

This prompted me to make this issue, if `const randomx::Program& getProgram()` could be added to main repo of randomx. This way everyone can access the `random_vm` programs in their codebases, experiments or tests without manually patching `virtual_machine.hpp`. 

I think that question "What program was used to calculate pow for the given block" will be often ask. At least this was the first question on my mined when I looked at the randomx.
